src/main/java/tech/stackable/hadoop/StackableTopologyProvider.java

```diff
@@ -293,6 +317,102 @@ public List<String> resolve(List<String> names) {
     return result;
   }
 
+  /**
+   * If the names include listeners then these must be resolved against the dataNode IPs and used
+   * subsequently.
+   *
+   * @param names the collection of names to resolve
+   * @param datanodes the dataNodes against which to conduct the resolution
+   * @return a collection of either the name (for non-listener) or the dataNode IP to which this
+   *     listener resolves
+   */
+  private List<String> dataNodesResolvedFromListenerOrOriginal(
+      List<String> names, List<Pod> datanodes) {
```

Comment:
I think it could be a good idea to add some tests around these methods:
- to make them easier to see what they're doing
- in case of future regressions

_same with `resolveDataNodesFromListener`, but I won't add a second comment for it_

Reply:
I've had a look at this and it doesn't appear to be straightforward due to the fact that client calls (difficult to mock) are often embedded behind cache checks, so passing them in to a method would mean executing calls that may be premature. I think there are some way around this e.g. using an abstract class or, even better, to implement a [pull through cache](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md#sharedinformers) as suggested by @nightkr, but I'd put these in a separate issue.

Reply:
Resolved as discussed. Pull through cache is added to issue https://github.com/stackabletech/hdfs-topology-provider/issues/6.